src/System/Process/Typed.hs

```diff
@@ -595,100 +626,104 @@ startProcess :: MonadIO m
              => ProcessConfig stdin stdout stderr
              -> m (Process stdin stdout stderr)
 startProcess pConfig'@ProcessConfig {..} = liftIO $ do
-    let cp0 =
-            case pcCmdSpec of
-                P.ShellCommand cmd -> P.shell cmd
-                P.RawCommand cmd args -> P.proc cmd args
-        cp = cp0
-            { P.std_in = ssStream pcStdin
-            , P.std_out = ssStream pcStdout
-            , P.std_err = ssStream pcStderr
-            , P.cwd = pcWorkingDir
-            , P.env = pcEnv
-            , P.close_fds = pcCloseFds
-            , P.create_group = pcCreateGroup
-            , P.delegate_ctlc = pcDelegateCtlc
+    ssStream pcStdin $ \realStdin ->
```

Comment:
There's something really subtle going on here: the cleanup code will be run as soon as the process is launched, not when the process finishes. Is that intended? It seems appropriate for the `/dev/null` case, but I'm not sure if it's generally applicable. (This may push back against my request to generalize the interface.)

Reply:
Correct, and for any `StreamSpec` created by `mkStreamSpec`, the function doesn't do any cleanup (it's just `($ ss)` over on line 496), so all existing code continues to behave in the same way.  It's only (right now) `nullStream` that passes in a function (line 523) that cleans up eagerly, since the handles it creates are not visible to calling code so there's no reason to keep them around.

When user code wants to eagerly clean up a handle it itself opened, it can do something like `withHandle $ \h -> startProcess (useHandleOpen h)`, so I don't think exposing `mkStreamSpec'` is necessary.  I'll remove the `since` and try to give it a better name, probably along the lines of `mkManagedStreamSpec`.